File: json_matcher.py
# ---- JSON Spec Matcher v2 -----------------------------------------
from typing import Any, Tuple, List, Dict
import json, re

class MatchError:
    __slots__ = ("path", "msg")
    def __init__(self, path: str, msg: str): self.path, self.msg = path, msg
    def __str__(self): return f"{self.path}: {self.msg}"

def _jsonpath(path: str, key: Any) -> str:
    if path == "$": 
        return f"$.{key}" if isinstance(key, str) else f"$[{key}]"
    return f"{path}.{key}" if isinstance(key, str) else f"{path}[{key}]"

def _typename(x: Any) -> str:
    if x is None: return "null"
    if isinstance(x, bool): return "boolean"
    if isinstance(x, (int, float)) and not isinstance(x, bool): return "number"
    if isinstance(x, str): return "string"
    if isinstance(x, list): return "array"
    if isinstance(x, dict): return "object"
    return type(x).__name__

def _is_op_dict(d: Any) -> bool:
    return isinstance(d, dict) and any(k.startswith("$") for k in d.keys())

def _match_scalar_op(spec: Dict[str, Any], actual: Any, path: str, errs: List[MatchError]) -> None:
    for op, val in spec.items():
        if op in ("$len","$any","$all","$subset","$items","$count","$strict","$forbid","$where"): 
            # handled elsewhere
            continue
        if op == "$eq":
            if actual != val:
                errs.append(MatchError(path, f"expected == {val!r}, got {actual!r}"))
        elif op == "$contains":
            if not isinstance(actual, str):
                errs.append(MatchError(path, f"$contains needs string, got {_typename(actual)}"))
            elif val not in actual:
                errs.append(MatchError(path, f"expected to contain {val!r}, got {actual!r}"))
        elif op == "$regex":
            if not isinstance(actual, str):
                errs.append(MatchError(path, f"$regex needs string, got {_typename(actual)}"))
            elif not re.search(val, actual):
                errs.append(MatchError(path, f"expected to match /{val}/, got {actual!r}"))
        elif op in ("$gt", "$ge", "$lt", "$le"):
            if not (isinstance(actual, (int, float)) and not isinstance(actual, bool)):
                errs.append(MatchError(path, f"{op} needs number, got {_typename(actual)}"))
            else:
                if op == "$gt" and not (actual > val): errs.append(MatchError(path, f"expected > {val}, got {actual}"))
                if op == "$ge" and not (actual >= val): errs.append(MatchError(path, f"expected >= {val}, got {actual}"))
                if op == "$lt" and not (actual < val): errs.append(MatchError(path, f"expected < {val}, got {actual}"))
                if op == "$le" and not (actual <= val): errs.append(MatchError(path, f"expected <= {val}, got {actual}"))
        elif op == "$in":
            if actual not in val:
                errs.append(MatchError(path, f"expected one of {val!r}, got {actual!r}"))
        elif op == "$type":
            if _typename(actual) != val:
                errs.append(MatchError(path, f"expected type {val}, got {_typename(actual)}"))
        elif op == "$present":
            # handled at object level; presence implies success
            pass
        elif op == "$not":
            sub_errs: List[MatchError] = []
            _match_any(val, actual, path, sub_errs)
            if not sub_errs:
                errs.append(MatchError(path, f"$not failed: value matches forbidden spec"))
        elif op == "$oneOf":
            all_errs = []
            for i, s in enumerate(val):
                sub_errs: List[MatchError] = []
                _match_any(s, actual, path, sub_errs)
                if not sub_errs:
                    break
                all_errs.append((i, sub_errs))
            else:
                flat = "; ".join(f"alt[{i}] first error: {e}" for i, es in all_errs for e in es[:1])
                errs.append(MatchError(path, f"none of $oneOf matched; {flat}"))
        else:
            errs.append(MatchError(path, f"unknown operator {op}"))

def _match_object(spec: Dict[str, Any], actual: Any, path: str, errs: List[MatchError]) -> None:
    if not isinstance(actual, dict):
        errs.append(MatchError(path, f"expected object, got {_typename(actual)}"))
        return

    # strictness and forbid lists
    strict = bool(spec.get("$strict", False))
    forbid = set(spec.get("$forbid", [])) if isinstance(spec.get("$forbid", []), list) else set()

    # normal keys to check
    spec_keys = [k for k in spec.keys() if not k.startswith("$")]

    if strict:
        extras = set(actual.keys()) - set(spec_keys)
        if extras:
            errs.append(MatchError(path, f"$strict: unexpected keys present: {sorted(extras)}"))

    if forbid:
        present = sorted([k for k in forbid if k in actual])
        if present:
            errs.append(MatchError(path, f"$forbid: keys must be absent: {present}"))

    for k in spec_keys:
        if k not in actual:
            v = spec[k]
            if isinstance(v, dict) and v.get("$present") is True:
                continue
            errs.append(MatchError(_jsonpath(path, k), "missing key"))
            continue
        _match_any(spec[k], actual[k], _jsonpath(path, k), errs)

def _matches(want: Any, got: Any) -> Tuple[bool, List[MatchError]]:
    e: List[MatchError] = []
    _match_any(want, got, "$", e)
    return (len(e) == 0), e

def _array_count(where_spec: Any, actual: List[Any], path: str, cmp_ops: Dict[str, Any], errs: List[MatchError]) -> None:
    n = 0
    for i, el in enumerate(actual):
        ok, _ = _matches(where_spec, el)
        if ok: n += 1
    _match_scalar_op(cmp_ops, n, path + "/$count", errs)

def _unordered_multiset_match(spec_items: List[Any], actual_items: List[Any], path: str, errs: List[MatchError]) -> None:
    # backtracking matcher for better pairing than greedy
    used = [False] * len(actual_items)
    best_hint = None  # store a first-failure hint

    def backtrack(j: int) -> bool:
        nonlocal best_hint
        if j == len(spec_items):
            return True
        want = spec_items[j]
        for i, got in enumerate(actual_items):
            if used[i]: 
                continue
            sub_errs: List[MatchError] = []
            _match_any(want, got, _jsonpath(path, i), sub_errs)
            if not sub_errs:
                used[i] = True
                if backtrack(j + 1): 
                    return True
                used[i] = False
            elif best_hint is None:
                best_hint = sub_errs[0]
        return False

    if not backtrack(0):
        errs.append(MatchError(path, f"unordered multiset match failed"))
        if best_hint:
            errs.append(best_hint)

def _match_array_controller(spec: Dict[str, Any], actual: Any, path: str, errs: List[MatchError]) -> None:
    if not isinstance(actual, list):
        errs.append(MatchError(path, f"expected array, got {_typename(actual)}"))
        return

    if "$len" in spec:
        want = spec["$len"]
        if isinstance(want, dict):
            _match_scalar_op(want, len(actual), path + "/$len", errs)
        else:
            if len(actual) != want:
                errs.append(MatchError(path, f"expected length {want}, got {len(actual)}"))

    if "$count" in spec:
        payload = spec["$count"]
        where = payload.get("$where", {})
        cmp_ops = {k: v for k, v in payload.items() if k.startswith("$") and k not in ("$where",)}
        _array_count(where, actual, path, cmp_ops, errs)

    if "$any" in spec:
        each = spec["$any"]
        for i, el in enumerate(actual):
            sub_errs: List[MatchError] = []
            _match_any(each, el, _jsonpath(path, i), sub_errs)
            if not sub_errs:
                break
        else:
            errs.append(MatchError(path, f"$any failed: no element matched"))

    if "$all" in spec:
        each = spec["$all"]
        for i, el in enumerate(actual):
            _match_any(each, el, _jsonpath(path, i), errs)

    if "$subset" in spec:
        subset_specs = list(spec["$subset"])
        _unordered_multiset_match(subset_specs, actual, path, errs)

    if "$items" in spec:
        want_items = list(spec["$items"])
        order = spec.get("$order", "any")
        if order == "exact":
            if len(want_items) != len(actual):
                errs.append(MatchError(path, f"$items exact: length mismatch {len(want_items)} != {len(actual)}"))
            else:
                for i, (w, a) in enumerate(zip(want_items, actual)):
                    _match_any(w, a, _jsonpath(path, i), errs)
        else:
            if len(want_items) != len(actual):
                errs.append(MatchError(path, f"$items any: actual has {len(actual)} items, spec has {len(want_items)}"))
            _unordered_multiset_match(want_items, actual, path, errs)

def _match_list(spec: Any, actual: Any, path: str, errs: List[MatchError]) -> None:
    if not isinstance(actual, list):
        errs.append(MatchError(path, f"expected array, got {_typename(actual)}"))
        return
    _unordered_multiset_match(list(spec), actual, path, errs)

def _match_any(spec: Any, actual: Any, path: str, errs: List[MatchError]) -> None:
    if isinstance(spec, dict) and any(k in spec for k in ("$len","$any","$all","$subset","$items","$count")):
        _match_array_controller(spec, actual, path, errs)
        return
    if _is_op_dict(spec) and not any(k in spec for k in ("$len","$any","$all","$subset","$items","$count")):
        _match_scalar_op(spec, actual, path, errs)
        return
    if isinstance(spec, dict):
        _match_object(spec, actual, path, errs)
        return
    if isinstance(spec, list):
        _match_list(spec, actual, path, errs)
        return
    if spec != actual:
        errs.append(MatchError(path, f"expected {spec!r}, got {actual!r}"))

def _pretty_json(s: str) -> str:
    try:
        return json.dumps(json.loads(s), indent=2, ensure_ascii=False, sort_keys=True)
    except Exception:
        return s

def validate_json_spec(actual_json_text: str, spec_json_text: str) -> Tuple[bool, str, str]:
    try:
        actual = json.loads(actual_json_text)
    except Exception as e:
        return False, f"Actual errors-file is not valid JSON: {e}", _pretty_json(actual_json_text)
    try:
        spec = json.loads(spec_json_text)
    except Exception as e:
        return False, f"Spec file is not valid JSON: {e}", _pretty_json(actual_json_text)

    errs: List[MatchError] = []
    _match_any(spec, actual, "$", errs)
    if not errs:
        return True, "OK", json.dumps(actual, indent=2, ensure_ascii=False, sort_keys=True)

    lines = ["JSON spec mismatches:"]
    for e in errs:
        lines.append(f"  - {e}")
    lines.append("\nActual JSON:")
    lines.append(json.dumps(actual, indent=2, ensure_ascii=False, sort_keys=True))
    return False, "\n".join(lines), json.dumps(actual, indent=2, ensure_ascii=False, sort_keys=True)
# ---- end JSON Spec Matcher v2 -------------------------------------

File: test_modules/json_matcher.ts
// test_modules/json_matcher.ts

type Json = null | boolean | number | string | Json[] | { [key: string]: Json };

class MatchError {
    path: string;
    msg: string;

    constructor(path: string, msg: string) {
        this.path = path;
        this.msg = msg;
    }

    toString(): string {
        return `${this.path}: ${this.msg}`;
    }
}

function jsonPath(path: string, key: string | number): string {
    if (path === "$") {
        return typeof key === "string" ? `$.${key}` : `$[${key}]`;
    }
    return typeof key === "string" ? `${path}.${key}` : `${path}[${key}]`;
}

function typeName(x: unknown): string {
    if (x === null) {
        return "null";
    }
    if (typeof x === "boolean") {
        return "boolean";
    }
    if (typeof x === "number") {
        return "number";
    }
    if (typeof x === "string") {
        return "string";
    }
    if (Array.isArray(x)) {
        return "array";
    }
    if (typeof x === "object") {
        return "object";
    }
    return typeof x;
}

function isOpDict(d: unknown): d is Record<string, Json> {
    if (!d || typeof d !== "object") {
        return false;
    }
    for (const k of Object.keys(d as Record<string, Json>)) {
        if (k.startsWith("$")) {
            return true;
        }
    }
    return false;
}

// forward declarations
function matchAny(spec: Json, actual: Json, path: string, errs: MatchError[]): void;

function matchAny(spec: Json, actual: Json, path: string, errs: MatchError[]): void {
    if (typeof spec === "object" && spec !== null && !Array.isArray(spec)) {
        const specObj = spec as Record<string, Json>;
        const hasArrayOps = ["$len", "$any", "$all", "$subset", "$items", "$count"].some((k) =>
            Object.prototype.hasOwnProperty.call(specObj, k)
        );
        if (hasArrayOps) {
            matchArrayController(specObj, actual, path, errs);
            return;
        }
        if (isOpDict(spec)) {
            matchScalarOp(specObj, actual, path, errs);
            return;
        }
        matchObject(specObj, actual, path, errs);
        return;
    }

    if (Array.isArray(spec)) {
        throw new Error("array spec must use array controller operators");
    }

    if (spec !== actual) {
        errs.push(new MatchError(path, `expected ${JSON.stringify(spec)}, got ${JSON.stringify(actual)}`));
    }
}

function matchScalarOp(spec: Record<string, Json>, actual: Json, path: string, errs: MatchError[]): void {
    for (const [op, val] of Object.entries(spec)) {
        if (["$len", "$any", "$all", "$subset", "$items", "$count", "$strict", "$forbid", "$where"].includes(op)) {
            continue;
        }

        if (op === "$eq") {
            if (actual !== val) {
                errs.push(new MatchError(path, `expected == ${JSON.stringify(val)}, got ${JSON.stringify(actual)}`));
            }
        } else if (op === "$contains") {
            if (typeof actual !== "string") {
                errs.push(new MatchError(path, `$contains needs string, got ${typeName(actual)}`));
            } else if (typeof val !== "string" || !actual.includes(val)) {
                errs.push(new MatchError(path, `expected to contain ${JSON.stringify(val)}, got ${JSON.stringify(actual)}`));
            }
        } else if (op === "$regex") {
            if (typeof actual !== "string") {
                errs.push(new MatchError(path, `$regex needs string, got ${typeName(actual)}`));
            } else {
                const re = new RegExp(String(val));
                if (!re.test(actual)) {
                    errs.push(new MatchError(path, `expected to match /${val}/, got ${JSON.stringify(actual)}`));
                }
            }
        } else if (op === "$gt" || op === "$ge" || op === "$lt" || op === "$le") {
            if (typeof actual !== "number") {
                errs.push(new MatchError(path, `${op} needs number, got ${typeName(actual)}`));
            } else {
                const numVal = Number(val);
                if (op === "$gt" && !(actual > numVal)) {
                    errs.push(new MatchError(path, `expected > ${numVal}, got ${actual}`));
                }
                if (op === "$ge" && !(actual >= numVal)) {
                    errs.push(new MatchError(path, `expected >= ${numVal}, got ${actual}`));
                }
                if (op === "$lt" && !(actual < numVal)) {
                    errs.push(new MatchError(path, `expected < ${numVal}, got ${actual}`));
                }
                if (op === "$le" && !(actual <= numVal)) {
                    errs.push(new MatchError(path, `expected <= ${numVal}, got ${actual}`));
                }
            }
        } else if (op === "$in") {
            if (!Array.isArray(val)) {
                errs.push(new MatchError(path, `$in needs array, got ${typeName(val)}`));
            } else {
                let found = false;
                for (const candidate of val) {
                    if (candidate === actual) {
                        found = true;
                        break;
                    }
                }
                if (!found) {
                    errs.push(new MatchError(path, `expected one of ${JSON.stringify(val)}, got ${JSON.stringify(actual)}`));
                }
            }
        } else if (op === "$type") {
            if (typeName(actual) !== String(val)) {
                errs.push(new MatchError(path, `expected type ${val}, got ${typeName(actual)}`));
            }
        } else if (op === "$present") {
            // handled at object level
        } else if (op === "$not") {
            const subErrs: MatchError[] = [];
            matchAny(val as Json, actual, path, subErrs);
            if (subErrs.length === 0) {
                errs.push(new MatchError(path, `$not failed: value matches forbidden spec`));
            }
        } else if (op === "$oneOf") {
            if (!Array.isArray(val)) {
                errs.push(new MatchError(path, `$oneOf needs array, got ${typeName(val)}`));
                continue;
            }
            const allErrs: Array<[number, MatchError[]]> = [];
            let matched = false;
            for (let i = 0; i < val.length; i++) {
                const s = val[i] as Json;
                const subErrs: MatchError[] = [];
                matchAny(s, actual, path, subErrs);
                if (subErrs.length === 0) {
                    matched = true;
                    break;
                }
                allErrs.push([i, subErrs]);
            }
            if (!matched) {
                const flat = allErrs
                    .map(([i, es]) => `alt[${i}] first error: ${es[0].toString()}`)
                    .join("; ");
                errs.push(new MatchError(path, `none of $oneOf matched; ${flat}`));
            }
        } else {
            errs.push(new MatchError(path, `unknown operator ${op}`));
        }
    }
}

function matchObject(spec: Record<string, Json>, actual: Json, path: string, errs: MatchError[]): void {
    if (typeof actual !== "object" || actual === null || Array.isArray(actual)) {
        errs.push(new MatchError(path, `expected object, got ${typeName(actual)}`));
        return;
    }

    const actualObj = actual as Record<string, Json>;

    const strict = Boolean(spec["$strict"]);
    const forbidVal = spec["$forbid"];
    const forbid = Array.isArray(forbidVal) ? new Set<string>(forbidVal.map(String)) : new Set<string>();

    const specKeys = Object.keys(spec).filter((k) => !k.startsWith("$"));

    if (strict) {
        const extras = Object.keys(actualObj).filter((k) => !specKeys.includes(k));
        if (extras.length > 0) {
            const sorted = [...extras].sort();
            errs.push(new MatchError(path, `$strict: unexpected keys present: ${JSON.stringify(sorted)}`));
        }
    }

    if (forbid.size > 0) {
        const present = [...forbid].filter((k) => Object.prototype.hasOwnProperty.call(actualObj, k)).sort();
        if (present.length > 0) {
            errs.push(new MatchError(path, `$forbid: keys must be absent: ${JSON.stringify(present)}`));
        }
    }

    for (const k of specKeys) {
        if (!Object.prototype.hasOwnProperty.call(actualObj, k)) {
            const v = spec[k];
            if (typeof v === "object" && v !== null && !Array.isArray(v) && v["$present"] === true) {
                continue;
            }
            errs.push(new MatchError(jsonPath(path, k), "missing key"));
            continue;
        }
        matchAny(spec[k] as Json, actualObj[k], jsonPath(path, k), errs);
    }
}

function matches(want: Json, got: Json): [boolean, MatchError[]] {
    const e: MatchError[] = [];
    matchAny(want, got, "$", e);
    return [e.length === 0, e];
}

function arrayCount(whereSpec: Json, actual: Json, path: string, cmpOps: Record<string, Json>, errs: MatchError[]): void {
    if (!Array.isArray(actual)) {
        errs.push(new MatchError(path, `expected array for $count, got ${typeName(actual)}`));
        return;
    }
    let n = 0;
    for (let i = 0; i < actual.length; i++) {
        const el = actual[i] as Json;
        const [ok] = matches(whereSpec, el);
        if (ok) {
            n += 1;
        }
    }
    matchScalarOp(cmpOps, n as unknown as Json, `${path}/$count`, errs);
}

function unorderedKeyedMatch(
    specItems: Json[],
    actualItems: Json[],
    keySpec: Json,   // string | string[]
    path: string,
    errs: MatchError[]
): void {
    type Key = string;

    function extract(obj: Json, itemPath: string): Key | null {
        if (typeof obj !== "object" || obj === null || Array.isArray(obj)) {
            errs.push(new MatchError(itemPath, "$key requires object elements"));
            return null;
        }

        const o = obj as Record<string, Json>;

        if (typeof keySpec === "string") {
            return JSON.stringify([o[keySpec]]);
        }

        if (Array.isArray(keySpec)) {
            const out: Json[] = [];
            for (const f of keySpec) {
                if (typeof f !== "string") {
                    errs.push(new MatchError(itemPath, "$key fields must be strings"));
                    return null;
                }
                out.push(o[f]);
            }
            return JSON.stringify(out);
        }

        errs.push(new MatchError(path, "$key must be string or array of strings"));
        return null;
    }

    const specKeys: Key[] = [];
    const actualKeys: Key[] = [];

    const specKeysMap = new Map<Key, Json>();
    const actualKeysMap = new Map<Key, Json>();

    for (let i = 0; i < specItems.length; i++) {
        const it = specItems[i];
        const k = extract(it, jsonPath(path, i));
        if (k !== null) {
            specKeys.push(k);
            specKeysMap.set(k, it);
        }
    }

    for (let i = 0; i < actualItems.length; i++) {
        const it = actualItems[i];
        const k = extract(it, jsonPath(path, i));
        if (k !== null) {
            actualKeys.push(k);
            actualKeysMap.set(k, it);
        }
    }

    const missing: Key[] = [];
    const extra: Key[] = [];

    const actualKeysCopy = [...actualKeys];
    const specKeysCopy = [...specKeys];

    const remove_in_place = (arr: Key[], val: Key) => {
        const idx = arr.indexOf(val);
        if (idx >= 0) {
            arr.splice(idx, 1);
        }
    }

    for (const k of specKeys) {
        const match = actualKeysCopy.find((ak) => ak === k);
        if (!match) {
            missing.push(k);
        } else {
            remove_in_place(actualKeysCopy, k);
        }
    }

    for (const k of actualKeys) {
        const match = specKeysCopy.find((sk) => sk === k);
        if (!match) {
            extra.push(k);
        } else {
            remove_in_place(specKeysCopy, k);
        }
    }

    if (missing.length > 0) {
        errs.push(new MatchError(path, `missing keys: ${missing.join(", ")}`));
    }

    if (extra.length > 0) {
        errs.push(new MatchError(path, `extra keys: ${extra.join(", ")}`));
    }

    if (errs.length > 0) {
        return;
    }

    // Same keys → deep match items
    for (const k of specKeys) {
        const want = specKeysMap.get(k);
        const got = actualKeysMap.get(k);

        if (want == null || got == null) {
            throw new Error("unreachable");
        }

        matchAny(want, got, `${path}{key=${k}}`, errs);
    }
}


function matchArrayController(spec: Record<string, Json>, actual: Json, path: string, errs: MatchError[]): void {
    if (!Array.isArray(actual)) {
        errs.push(new MatchError(path, `expected array, got ${typeName(actual)}`));
        return;
    }

    if ("$len" in spec) {
        const want = spec["$len"] as Json;
        if (typeof want === "object" && want !== null && !Array.isArray(want)) {
            matchScalarOp(want as Record<string, Json>, actual.length as unknown as Json, `${path}/$len`, errs);
        } else {
            const n = Number(want);
            if (actual.length !== n) {
                errs.push(new MatchError(path, `expected length ${n}, got ${actual.length}`));
            }
        }
    }

    if ("$count" in spec) {
        const payload = spec["$count"];
        if (typeof payload === "object" && payload !== null && !Array.isArray(payload)) {
            const payloadObj = payload as Record<string, Json>;
            const where = (payloadObj["$where"] ?? {}) as Json;
            const cmpOps: Record<string, Json> = {};
            for (const [k, v] of Object.entries(payloadObj)) {
                if (k.startsWith("$") && k !== "$where") {
                    cmpOps[k] = v;
                }
            }
            arrayCount(where, actual, path, cmpOps, errs);
        }
    }

    if ("$any" in spec) {
        const each = spec["$any"] as Json;
        let okAny = false;
        for (let i = 0; i < actual.length; i++) {
            const el = actual[i] as Json;
            const subErrs: MatchError[] = [];
            matchAny(each, el, jsonPath(path, i), subErrs);
            if (subErrs.length === 0) {
                okAny = true;
                break;
            }
        }
        if (!okAny) {
            errs.push(new MatchError(path, "$any failed: no element matched"));
        }
    }

    if ("$all" in spec) {
        const each = spec["$all"] as Json;
        for (let i = 0; i < actual.length; i++) {
            const el = actual[i] as Json;
            matchAny(each, el, jsonPath(path, i), errs);
        }
    }

    if ("$subset" in spec) {
        const subsetSpecs = Array.isArray(spec["$subset"]) ? spec["$subset"] as Json[] : [];
        const keySpec = spec["$key"];
        unorderedKeyedMatch(subsetSpecs, actual as Json[], keySpec, path, errs);
    }

    if ("$items" in spec) {
        const wantItems = Array.isArray(spec["$items"]) ? spec["$items"] as Json[] : [];
        const order = (spec["$order"] as string) ?? "any";
        if (order === "exact") {
            if (wantItems.length !== actual.length) {
                errs.push(new MatchError(path, `$items exact: length mismatch ${wantItems.length} != ${actual.length}`));
            } else {
                for (let i = 0; i < wantItems.length; i++) {
                    matchAny(wantItems[i], actual[i] as Json, jsonPath(path, i), errs);
                }
            }
        } else {
            if (wantItems.length !== actual.length) {
                errs.push(new MatchError(path, `$items any: actual has ${actual.length} items, spec has ${wantItems.length}`));
            }
            const keySpec = spec["$key"];
            unorderedKeyedMatch(wantItems, actual as Json[], keySpec, path, errs);
        }
    }
}

function prettyJson(s: string): string {
    try {
        const parsed = JSON.parse(s);
        return JSON.stringify(parsed, null, 2);
    } catch {
        return s;
    }
}

export function validateJsonSpec(
    actualJsonText: string,
    specJsonText: string
): [boolean, string, string] {
    let actual: Json;
    try {
        actual = JSON.parse(actualJsonText) as Json;
    } catch (e) {
        return [false, `Actual errors-file is not valid JSON: ${String(e)}`, prettyJson(actualJsonText)];
    }

    let spec: Json;
    try {
        spec = JSON.parse(specJsonText) as Json;
    } catch (e) {
        return [false, `Spec file is not valid JSON: ${String(e)}`, prettyJson(actualJsonText)];
    }

    const errs: MatchError[] = [];
    matchAny(spec, actual, "$", errs);

    if (errs.length === 0) {
        return [true, "OK", JSON.stringify(actual, null, 2)];
    }

    const lines: string[] = ["JSON spec mismatches:"];
    for (const e of errs) {
        lines.push(`  - ${e.toString()}`);
    }
    return [false, lines.join("\n"), JSON.stringify(actual, null, 2)];
}
